File: medium/1914-cyclically-rotating-a-grid/solution.py
from typing import List


class Solution:
    def rotateGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        """
        1914. Cyclically Rotating a Grid
        Time: O(m × n)
        Space: O(m + n)
        """
        m, n = len(grid), len(grid[0])
        layers = min(m, n) // 2

        for layer in range(layers):
            coords = self._layer_coordinates(m, n, layer)
            values = [grid[row][col] for row, col in coords]
            shift = k % len(values)

            for i, (row, col) in enumerate(coords):
                grid[row][col] = values[(i - shift) % len(values)]

        return grid

    def _layer_coordinates(self, m: int, n: int, layer: int) -> List[tuple[int, int]]:
        top = layer
        bottom = m - 1 - layer
        left = layer
        right = n - 1 - layer
        coords = []

        for row in range(top, bottom + 1):
            coords.append((row, left))

        for col in range(left + 1, right + 1):
            coords.append((bottom, col))

        for row in range(bottom - 1, top - 1, -1):
            coords.append((row, right))

        for col in range(right - 1, left, -1):
            coords.append((top, col))

        return coords

File: medium/1914-cyclically-rotating-a-grid/solution.java
import java.util.*;

/**
 * 1914. Cyclically Rotating a Grid
 * Time: O(m × n)
 * Space: O(m + n)
 */
class Solution {
    public int[][] rotateGrid(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        int layers = Math.min(m, n) / 2;

        for (int layer = 0; layer < layers; layer++) {
            List<int[]> coords = layerCoordinates(m, n, layer);
            int size = coords.size();
            int[] values = new int[size];

            for (int i = 0; i < size; i++) {
                int[] cell = coords.get(i);
                values[i] = grid[cell[0]][cell[1]];
            }

            int shift = k % size;

            for (int i = 0; i < size; i++) {
                int[] cell = coords.get(i);
                grid[cell[0]][cell[1]] = values[(i - shift + size) % size];
            }
        }

        return grid;
    }

    private List<int[]> layerCoordinates(int m, int n, int layer) {
        int top = layer;
        int bottom = m - 1 - layer;
        int left = layer;
        int right = n - 1 - layer;
        List<int[]> coords = new ArrayList<>();

        for (int row = top; row <= bottom; row++) {
            coords.add(new int[]{row, left});
        }

        for (int col = left + 1; col <= right; col++) {
            coords.add(new int[]{bottom, col});
        }

        for (int row = bottom - 1; row >= top; row--) {
            coords.add(new int[]{row, right});
        }

        for (int col = right - 1; col > left; col--) {
            coords.add(new int[]{top, col});
        }

        return coords;
    }
}
